shared/pubsub.js

```diff
@@ -0,0 +1,608 @@
+  },
+
+  // Schedules a connection attempt to happen after a delay computed according to
+  // a randomized exponential backoff algorithm variant.
+  scheduleConnectionAttempt () {
+    const client = this
+
+    if (!client.shouldReconnect) {
+      throw new Error('Cannot call `scheduleConnectionAttempt()` when `shouldReconnect` is false.')
+    }
+    const delay = client.getNextRandomDelay()
+    const nth = client.failedConnectionAttempts + 1
+
+    client.nextConnectionAttemptDelayID = setTimeout(() => {
+      sbp('okTurtles.events/emit', PUBSUB_RECONNECTION_ATTEMPT, client)
+      client.nextConnectionAttemptDelayID = undefined
+      client.connect()
+    }, delay)
+    sbp('okTurtles.events/emit', PUBSUB_RECONNECTION_SCHEDULED, client, { delay, nth })
+  },
+
+  // Unused for now.
+  pub (contractID: string, data: JSONType) {
+  },
+
+  /**
+   * Sends a SUB request to the server as soon as possible.
+   *
+   * - The given contract ID will be cached until we get a relevant server
+   * response, allowing us to resend the same request if necessary.
+   * - Any identical UNSUB request that has not been sent yet will be cancelled.
+   * - Calling this method again before the server has responded has no effect.
+   * @param contractID - The ID of the contract whose updates we want to subscribe to.
+   */
+  sub (contractID: string) {
+    const client = this
+    const { socket } = this
+
+    if (!client.pendingSubscriptionSet.has(contractID)) {
+      if (socket && socket.readyState === WebSocket.OPEN) {
+        socket.send(createRequest(REQUEST_TYPE.SUB, { contractID }))
+      }
+    }
+    client.pendingSubscriptionSet.add(contractID)
+    client.pendingUnsubscriptionSet.delete(contractID)
+  },
+
+  /**
+   * Sends an UNSUB request to the server as soon as possible.
+   *
+   * - The given contract ID will be cached until we get a relevant server
+   * response, allowing us to resend the same request if necessary.
+   * - Any identical SUB request that has not been sent yet will be cancelled.
+   * - Calling this method again before the server has responded has no effect.
+   * @param contractID - The ID of the contract whose updates we want to unsubscribe from.
+   */
+  unsub (contractID: string) {
+    const client = this
+    const { socket } = this
+
+    if (!client.pendingUnsubscriptionSet.has(contractID)) {
+      if (socket && socket.readyState === WebSocket.OPEN) {
+        socket.send(createRequest(REQUEST_TYPE.UNSUB, { contractID }))
+      }
+    }
+    client.pendingSubscriptionSet.delete(contractID)
+    client.pendingUnsubscriptionSet.add(contractID)
+  }
+}
+
+// Register custom SBP event listeners before the first connection.
+for (const name of Object.keys(defaultClientEventHandlers)) {
+  if (name === 'error' || !socketEventNames.includes(name)) {
+    sbp('okTurtles.events/on', `pubsub-${name}`, (target, detail?: Object) => {
+      target.listeners[name]({ type: name, target, detail })
+    })
```

Comment:
Wait... this doesn't look right to me. This appears to be re-implementing an event system ... using an event system.

Reply:
Copying my reply from Slack:

```
 FYI, I think it might update #dev, but it seems like sometimes it drops the messages. Anyway, this might be a better place to discuss it because it's faster and when there's a lot of back & forth Slack is better suited to it
5:45
I got a bit distracted by something but now am looking closer at it. I now see clearer what's going on... it's really kind of a conflict between two different approaches / coding styles... I guess what you have here is OK. It just looks weird because the okTurtles.events API is meant to be a more global events system, and here you have handlers for a specific instance of a pubsub client, and only a single handler per instance. A very different sort of approach
5:46
but it makes sense
5:47
In fact, we can keep this weirdness around to remind us to in the future maybe convert the API to a more SBP-style approach (as a possibility / experiment, not a necessity)
```

Feel free to continue the conversation there. I'll resolve this here in the meantime.